File: PowerShellAutomationAgent.UI/src/services/projects/project.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { PaginatedList } from '../pagination/paginated-list';
import { ProjectResponseModel } from './models/project-response-model';
import { AddProjectRequestModel } from './models/add-project-request-model';
import { AddProjectResponseModel } from './models/add-project-response-model';
import { EditProjectRequestModel } from './models/edit-project-request-model';

@Injectable({
  providedIn: 'root'
})
export class ProjectService {

  private apiUrl = '/api/project';

  constructor(private http: HttpClient) { }

  getProjects(pageSize = 2147483647, page = 0): Observable<PaginatedList<ProjectResponseModel>> {
    return this.http.get<PaginatedList<ProjectResponseModel>>(`${this.apiUrl}?pageSize=${pageSize}&page=${page}`);
  }

  getProject(projectId: string) {
    return this.http.get<ProjectResponseModel>(`${this.apiUrl}/${projectId}`);
  }

  addProject(addProjectRequestModel: AddProjectRequestModel): Observable<AddProjectResponseModel> {
    return this.http.post<AddProjectResponseModel>(`${this.apiUrl}`, addProjectRequestModel);
  }

  editProject(projectId: string, editProjectRequestModel: EditProjectRequestModel) {
    return this.http.put(`${this.apiUrl}/${projectId}`, editProjectRequestModel);
  }

  deleteProject(projectId: string) {
    return this.http.delete(`${this.apiUrl}/${projectId}`);
  }
}

File: PowerShellAutomationAgent.API/PowerShellAutomationAgent.API/Services/Projects/ProjectService.cs
using PowerShellAutomationAgent.API.Data;
using PowerShellAutomationAgent.API.Services.Pagination;
using PowerShellAutomationAgent.API.Services.Projects.Models;
using Microsoft.EntityFrameworkCore;

namespace PowerShellAutomationAgent.API.Services.Projects;

public class ProjectService(DatabaseContext db) : IProjectService
{
    public async Task<AddProjectResponseModel> AddProjects(AddProjectRequestModel requestModel)
    {
        var project = new Data.Models.Project()
        {
            Id = Guid.NewGuid(),
            Name = requestModel.Name,
            CreatedOnUtc = DateTimeOffset.UtcNow
        };

        db.Projects.Add(project);
        await db.SaveChangesAsync();

        return new AddProjectResponseModel()
        {
            Id = project.Id
        };
    }

    public async Task<bool> DeleteProject(Guid projectId)
    {
        var project = await db.Projects.Where(x => x.Id == projectId).FirstOrDefaultAsync();
        if (project == null)
        {
            return false;
        }

        db.Remove(project);
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> EditProject(Guid projectId, EditProjectRequestModel requestModel)
    {
        var project = await db.Projects.Where(x => x.Id == projectId).FirstOrDefaultAsync();
        if (project == null)
        {
            return false;
        }

        project.Name = requestModel.Name;
        project.Script = requestModel.Script;
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<ProjectResponseModel?> GetProject(Guid projectId)
    {
        return await db.Projects.AsNoTracking()
            .Where(x => x.Id == projectId)
            .Select(x => new ProjectResponseModel()
            {
                Id = x.Id,
                Name = x.Name,
                CreatedOnUtc = x.CreatedOnUtc,
                Script = x.Script,
            })
            .FirstOrDefaultAsync();
    }

    public async Task<PaginatedList<ProjectResponseModel>> GetProjects(int pageSize = int.MaxValue, int page = 0)
    {
        var data = await db.Projects.AsNoTracking()
            .OrderByDescending(x => x.CreatedOnUtc)
            .Skip(page * pageSize)
            .Take(pageSize)
            .Select(x => new ProjectResponseModel()
            {
                Id = x.Id,
                Name = x.Name,
                CreatedOnUtc = x.CreatedOnUtc
            })
            .ToListAsync();

        return new PaginatedList<ProjectResponseModel>()
        {
            Data = data,
            HasNext = await db.Projects.Skip((page + 1) * pageSize).AnyAsync(),
            HasPrevious = page > 0,
            Page = page,
            PageSize = pageSize
        };
    }
}
